cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -275,56 +280,127 @@ void accumulate_vertex_results(
+    [] __device__(auto, auto d) { return (d == invalid_distance) ? vertex_t{0} : vertex_t{1}; },
+    vertex_t{0},
+    reduce_op::plus<vertex_t>{},
+    do_expensive_check);
+  
+  rmm::device_uvector<vertex_t> reusable_vertex_buffer(max_frontier_size, handle.get_stream());
+  rmm::device_uvector<weight_t> reusable_delta_buffer(max_frontier_size, handle.get_stream());
+  rmm::device_uvector<weight_t> reusable_centrality_buffer(max_frontier_size, handle.get_stream());
+
   // Based on Brandes algorithm, we want to follow back pointers in non-increasing
   // distance from S to compute delta
   //
   for (vertex_t d = diameter; d > 1; --d) {
-    per_v_transform_reduce_outgoing_e(
+    // Clear deltas array for this iteration
+    detail::scalar_fill(handle, deltas.data(), deltas.size(), weight_t{0});
+    
+    // Find vertices at distance d-1 (frontier vertices)
+    auto frontier_count = count_if_v(
       handle,
       graph_view,
-      src_properties.view(),
-      dst_properties.view(),
-      cugraph::edge_dummy_property_t{}.view(),
-      [d] __device__(auto, auto, auto src_props, auto dst_props, auto) {
-        if ((thrust::get<0>(dst_props) == d) && (thrust::get<0>(src_props) == (d - 1))) {
-          auto sigma_v = static_cast<weight_t>(thrust::get<1>(src_props));
-          auto sigma_w = static_cast<weight_t>(thrust::get<1>(dst_props));
-          auto delta_w = thrust::get<2>(dst_props);
-
-          return (sigma_v / sigma_w) * (1 + delta_w);
-        } else {
-          return weight_t{0};
-        }
-      },
-      weight_t{0},
-      reduce_op::plus<weight_t>{},
-      deltas.begin(),
+      distances.begin(),
+      [d] __device__(auto, auto distance) { return distance == (d - 1); },
       do_expensive_check);
-
-    update_edge_src_property(
-      handle,
-      graph_view,
-      thrust::make_zip_iterator(distances.begin(), sigmas.begin(), deltas.begin()),
-      src_properties.mutable_view());
-    update_edge_dst_property(
-      handle,
-      graph_view,
-      thrust::make_zip_iterator(distances.begin(), sigmas.begin(), deltas.begin()),
-      dst_properties.mutable_view());
-
-    thrust::transform(handle.get_thrust_policy(),
-                      centralities.begin(),
-                      centralities.end(),
-                      deltas.begin(),
-                      centralities.begin(),
-                      thrust::plus<weight_t>());
+    
+    if (frontier_count > 0) {
+      // Extract frontier vertices
+      auto v_first = graph_view.local_vertex_partition_range_first();
+      thrust::copy_if(
+        handle.get_thrust_policy(),
+        thrust::make_counting_iterator(v_first),
+        thrust::make_counting_iterator(graph_view.local_vertex_partition_range_last()),
+        reusable_vertex_buffer.begin(),
+        [distances = distances.begin(), d, v_first] __device__(vertex_t vertex) {
+          auto offset = vertex - v_first;
+          return distances[offset] == (d - 1);
+        });
```

Comment:
This shouldn't be necessary if we sort the vertices by distances outside the loop.